File: agent-factory/src/llm.js
// OpenRouter-style model client. LIVE when OPENROUTER_API_KEY is set; otherwise
// a deterministic offline STUB that makes zero network calls and never spends.
// Either way it returns { text, usage:{input_tokens,output_tokens}, mode }.
import { loadEnv } from './util.js';

loadEnv();

function approxTokens(s) { return Math.max(1, Math.ceil((s || '').length / 4)); }

// Deterministic offline completion. It is intentionally simple — the point of the
// factory is the orchestration, not the model. It echoes task-shaped guidance so
// downstream skills can still produce real artifacts offline.
function stubComplete({ system, prompt, model }) {
  const input_tokens = approxTokens(system) + approxTokens(prompt);
  const seedText =
    `[stub:${model}] ` +
    `Handled prompt deterministically offline. ` +
    `Key points extracted: ${(prompt || '').split(/\s+/).slice(0, 12).join(' ')}...`;
  const output_tokens = approxTokens(seedText) + 40; // pretend a bit of reasoning
  return { text: seedText, usage: { input_tokens, output_tokens }, mode: 'stub' };
}

async function liveComplete({ system, prompt, model }) {
  const base = process.env.OPENROUTER_BASE_URL || 'https://openrouter.ai/api/v1';
  const res = await fetch(`${base}/chat/completions`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${process.env.OPENROUTER_API_KEY}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      model,
      messages: [
        ...(system ? [{ role: 'system', content: system }] : []),
        { role: 'user', content: prompt },
      ],
    }),
  });
  if (!res.ok) throw new Error(`OpenRouter ${res.status}: ${await res.text()}`);
  const data = await res.json();
  const text = data.choices?.[0]?.message?.content ?? '';
  const usage = {
    input_tokens: data.usage?.prompt_tokens ?? approxTokens(system) + approxTokens(prompt),
    output_tokens: data.usage?.completion_tokens ?? approxTokens(text),
  };
  return { text, usage, mode: 'live' };
}

export function isLive() {
  return Boolean(process.env.OPENROUTER_API_KEY && process.env.OPENROUTER_API_KEY.trim());
}

export async function complete(opts) {
  return isLive() ? liveComplete(opts) : stubComplete(opts);
}

export default { complete, isLive };

File: agent-factory/src/llm.py
"""OpenRouter-style model client — STUBBED OFFLINE by default.

If OPENROUTER_API_KEY is set in .env, `complete()` will attempt a real HTTP
call. With no key (the default), it returns a deterministic mock completion so
the entire factory runs end-to-end with no network and no spend.

Token accounting is real in both modes so the cost ledger is meaningful.
"""
from __future__ import annotations

import json
import time
import urllib.request
from dataclasses import dataclass

from common import load_env


@dataclass
class Completion:
    text: str
    model: str
    prompt_tokens: int
    completion_tokens: int
    stubbed: bool


def _estimate_tokens(text: str) -> int:
    # ~4 chars/token heuristic, good enough for ledger math.
    return max(1, len(text) // 4)


def complete(model: str, prompt: str, *, max_tokens: int = 1200,
             system: str = "") -> Completion:
    env = load_env()
    key = env.get("OPENROUTER_API_KEY")
    prompt_tokens = _estimate_tokens(system + prompt)

    if not key:
        # ---- OFFLINE MOCK PATH -------------------------------------------
        # Deterministic, model-flavored stub. The real *deliverable content*
        # for the business-model task is produced by the task handler itself;
        # this stub stands in for the raw LLM call so routing/cost is exercised.
        body = (
            f"[stubbed completion · model={model}]\n"
            f"Acknowledged request ({prompt_tokens} prompt tokens). "
            f"Produced structured output per task contract."
        )
        completion_tokens = min(max_tokens, _estimate_tokens(body) + 80)
        # Small realistic latency so concurrent workers actually share load and
        # the latency metrics are meaningful. Stronger models "think" longer.
        weight = {"strong": 0.45, "mid": 0.22}.get(model.split("/")[0], 0.10)
        time.sleep(weight + completion_tokens / 4000)
        return Completion(
            text=body,
            model=model,
            prompt_tokens=prompt_tokens,
            completion_tokens=completion_tokens,
            stubbed=True,
        )

    # ---- REAL PATH (only if the user opted in with a key) -----------------
    url = env.get("OPENROUTER_BASE_URL", "https://openrouter.ai/api/v1") + "/chat/completions"
    payload = {
        "model": model,
        "messages": ([{"role": "system", "content": system}] if system else [])
        + [{"role": "user", "content": prompt}],
        "max_tokens": max_tokens,
    }
    req = urllib.request.Request(
        url, data=json.dumps(payload).encode(),
        headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json"},
    )
    with urllib.request.urlopen(req, timeout=60) as resp:
        data = json.loads(resp.read())
    text = data["choices"][0]["message"]["content"]
    usage = data.get("usage", {})
    return Completion(
        text=text,
        model=model,
        prompt_tokens=usage.get("prompt_tokens", prompt_tokens),
        completion_tokens=usage.get("completion_tokens", _estimate_tokens(text)),
        stubbed=False,
    )
